### PR TITLE
Cc11 geolocation

### PR DESCRIPTION
**Pull request by bancho22**
Solves #11

### CODE REVIEW

File: src/hooks.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+const google = window.google
+
+const getCountrySlug = coords => {
+  const code = countryIso.get(coords.latitude, coords.longitude)
+  return countryData.countries[code].name
+}
+
+const setUpGeolocation = async ({coords, setGeolocationState, setSelectedCountries, setData}) => {
+  const slug = getCountrySlug(coords).toLowerCase()
+  console.log(slug)
```
Comment (bancho22) [inline src/hooks.js:19]: Whoops

File: public/index.html [hunk 2]
```diff
@@ -25,6 +25,8 @@
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
 
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css">
```
Comment (bence-toth) [inline public/index.html:28]: I don't know how I feel about this :see_no_evil:

File: src/hooks.js (hunk 1 above)
Comment (bence-toth) [reply src/hooks.js:19]: Yeah maybe let's get rid of that :slightly_smiling_face:

File: src/geolocationStates.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+const geolocationStates = {
```
Comment (bence-toth) [inline src/geolocationStates.js:1]: Let's wrap this in `Object.freeze()`

File: src/App.js [hunk 4]
```diff
@@ -128,7 +97,15 @@ const App = () => {
             7-day moving average of COVID-19 <br />deaths per million people
           </div>
         </h1>
-        <div id='chart'></div>
+        {
+          [geolocationStates.loaded, geolocationStates.disallowed].includes(geolocationState)
+          ? (
+            <div id='chart'></div>
```
Comment (bence-toth) [inline src/App.js:103]: If this is rendered conditionally, let's make `drawChart()` bail out if the element it's trying to render to does not exist

File: src/App.js [hunk 5]
```diff
@@ -128,7 +97,15 @@ const App = () => {
...
+          )
+          : (
+            <Fallback geolocationState={geolocationState} />
```
Comment (bence-toth) [inline src/App.js:106]: When the page loads with `geolocationStates.disallowed`, it kinda explodes

File: src/Fallback.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        <div className='spinner-border text-warning'></div>
+      )
+    }
+    {
+      geolocationState === geolocationStates.loading && (
+        <div className='spinner-border text-success'></div>
+      )
+    }
+    {
+      geolocationState === geolocationStates.disallowed && (
```
Comment (bence-toth) [inline src/Fallback.js:16]: I don't think we'll ever get here. If `geolocationState` is `geolocationStates.disallowed`, `<Fallback />` is not rendered.

File: public/index.html (hunk 2 above)
Comment (bence-toth) [reply public/index.html:28]: Maybe let's just grab the relevant styles from the bootstap stylesheet instead?

**Review by bence-toth — CHANGES_REQUESTED** [review]

File: src/hooks.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  useLayoutEffect(() => {
+    if (google.visualization) {
+      if (!selectedCountries.some(country => !data[country])) {
+        drawChart({data, selectedCountries})
+      }
+    }
+  }, [data, selectedCountries])
+}
+
+const useResizeListener = ({data, selectedCountries}) => {
```
Comment (bence-toth) [inline src/hooks.js:52]: I noticed that the `window.resize` event handler is a bit trigger happy and fails if you resize the window before the application properly loads.

File: src/App.js (hunk 5 above)
Comment (bancho22) [reply src/App.js:106]: Hadn't thought it out well. Should make more sense now.

File: src/hooks.js (hunk 7 above)
Comment (bancho22) [reply src/hooks.js]: True. I added a new condition there. Not sure if it's the best way to approach it but it seems to work.

File: src/index.css [hunk 8]
```diff
@@ -100,6 +100,43 @@ h1 br {
...
+  to { transform: rotate(360deg); }
+}
+
+.spinner-border {
+  display: inline-block;
+  width: 2rem;
+  height: 2rem;
+  vertical-align: text-bottom;
+  border: .25em solid currentColor;
+      border-right-color: currentcolor;
```
Comment (bence-toth) [inline src/index.css:133]: Let's fix the indentation here:
```suggestion
  border-right-color: currentcolor;
```

File: src/index.css [hunk 9]
```diff
@@ -100,6 +100,43 @@ h1 br {
...
+  border-right-color: transparent;
+  border-radius: 50%;
+  -webkit-animation: spinner-border .75s linear infinite;
+  animation: spinner-border .75s linear infinite;
```
Comment (bence-toth) [inline src/index.css:137]: Let's get rid of `-webkit-animation`:
```suggestion
  animation: spinner-border .75s linear infinite;
```

**Review by bence-toth — CHANGES_REQUESTED** [review]

File: src/hooks.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  setGeolocationState(geolocationStates.loading)
+  google.charts.setOnLoadCallback(async () => {
+    const localData = await getCountryData(slug)
+    setGeolocationState(geolocationStates.loaded)
+    setSelectedCountries([slug])
+    setData({ [slug]: localData })
+  })
+}
+
+const handleGeoError = async ({code, fallbackCountry, setGeolocationState, setSelectedCountries, setData}) => {
```
Comment (bence-toth) [inline src/hooks.js:28]: If the Geolocation access is denied by the browser on page load, the application crashes with:
```
TypeError: google.visualization.arrayToDataTable is not a function
```

**Comment by bancho22** [comment]
Can we squash the commits?

File: src/hooks.js (hunk 10 above)
Comment (bancho22) [reply src/hooks.js:28]: I don't seem to be able to replicate that.

Comment (bence-toth) [reply src/hooks.js:28]: Let me try again

Comment (bence-toth) [reply src/hooks.js:28]: I can still reproduce it like this:
- Start the dev server
- Set the Geolocation access to "Ask"
- Reload the page
- When it asks for Geolocation access, hit "Allow"
- Set Geolocation access to "Deny" manually
- Reload the page

Comment (bence-toth) [reply src/hooks.js:28]: Interestingly I get a different error (`TypeError: google.visualization.LineChart is not a constructor`) with these steps:
- Start the dev server
- Set the Geolocation access to "Ask"
- Reload the page
- When it asks for Geolocation access, hit "Block"
- Reload the page

Comment (bence-toth) [reply src/hooks.js:28]: It looks like `drawChart` is called before `google.charts.setOnLoadCallback` is triggered.

Maybe it would make sense to create a totally separate hook for the loading state of the Google charts which has only a `useState` initialized to false and set to true on `google.charts.setOnLoadCallback`.

Comment (bancho22) [reply src/hooks.js:28]: Turns out the problem was that I wasn't calling `google.charts.setOnLoadCallback` at call in the `handleGeoError` function.

Should be fixed. Please verify and resolve.

**Review by bence-toth — APPROVED** [review]
:100: 

Thank you for this @bancho22! :tada:
